addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/TestPersistentXSS3FindPossibleSinks.java

```diff
@@ -0,0 +1,132 @@
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2013 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.ascanrules;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.apache.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.AbstractAppPlugin;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.core.scanner.Category;
+import org.parosproxy.paros.network.HttpMessage;
+
+public class TestPersistentXSS3FindPossibleSinks extends AbstractAppPlugin {
+
+    /** Prefix for internationalised messages used by this rule */
+    private static final String MESSAGE_PREFIX = "ascanrules.testpersistentxss3findpossiblesinks.";
+
+    private static Logger log = Logger.getLogger(TestPersistentXSS3FindPossibleSinks.class);
+
+    @Override
+    public int getId() {
+        return 40039;
+    }
+
+    @Override
+    public String getName() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "name");
+    }
+
+    @Override
+    public String[] getDependency() {
+        return new String[] {"TestPersistentXSS2ResendWithOriginalValues"};
+    }
+
+    @Override
+    public String getDescription() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "misc");
+    }
+
+    @Override
+    public int getCategory() {
+        return Category.INJECTION;
+    }
+
+    @Override
+    public String getSolution() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "misc");
+    }
+
+    @Override
+    public String getReference() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "misc");
+    }
+
+    @Override
+    public void scan() {
+        HttpMessage msg = getBaseMsg();
+        try {
+            HttpMessage msg1 = msg.cloneRequest();
+            sendAndReceive(msg1, false);
+            findPossibleSinks(msg1);
+
+        } catch (Exception e) {
+            log.error(e.getMessage(), e);
+        }
+    }
+
+    private void findPossibleSinks(HttpMessage msg) {
+        PersistentXSSStorage storage;
+        Object obj = getKb().get("XSSStorage");
+        if (obj instanceof PersistentXSSStorage) {
+            storage = (PersistentXSSStorage) obj;
+        } else {
+            return;
+        }
+
+        // the getKeysContainedIn returns repeated values
```

Comment:
Is this just meant to clarify why Set is used? If so I think it can probably be removed.

If it's a TODO meant to cause someone to sort out why there are repeated values then it should be noted as such.

Reply:
Yes it was to clarify that I was using the set to remove duplicated values.

Reply:
The method could just return a set to start with?

Reply:
No :/. What I will do is to abstract this inside the class PersistentXSSStorage